meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientInterface.java

```diff
@@ -20,139 +20,382 @@
       AsyncCallback.VoidCallback cb);
 
-  void asyncSet(final String key, T data, int version, AsyncCallback.VoidCallback cb);
+  /**
+   * The asynchronous version of set.
+   * @param key key to identify the entry
+   * @param data new data of the entry
+   * @param version expected version if the entry. -1 matched any version
+   * @param cb An user defined VoidCallback implementation that will be invoked when async create return.
+   */
+  void asyncSet(final String key, final T data, final int version, AsyncCallback.VoidCallback cb);
 
-  void asyncUpdate(final String key, DataUpdater<T> updater, AsyncCallback.VoidCallback cb);
+  /**
+   * The asynchronous version of update.
+   * @param key key to identify the entry
+   * @param updater An updater that modifies the entry value.
+   * @param cb An user defined VoidCallback implementation that will be invoked when async create return.
+   *           It will contain the newly updated data if update succeeded.
+   */
+  void asyncUpdate(final String key, DataUpdater<T> updater, AsyncCallback.DataCallback cb);
 
+  /**
+   * The asynchronous version of get.
+   * @param key key to identify the entry
+   * @param cb An user defined VoidCallback implementation that will be invoked when async create return.
+   *           It will contain the entry data if get succeeded.
+   */
   void asyncGet(final String key, AsyncCallback.DataCallback cb);
 
-  void asyncCountSubEntries(final String path, AsyncCallback.DataCallback cb);
+  /**
+   * The asynchronous version of get sub entries.
+   * @param key key to identify the entry
+   * @param cb An user defined VoidCallback implementation that will be invoked when async create return.
+   *           It will contain the list of sub entry keys if succeeded.
+   */
+  void asyncCountSubEntries(final String key, AsyncCallback.DataCallback cb);
 
+  /**
+   * The asynchronous version of get sub entries.
+   * @param key key to identify the entry
+   * @param cb An user defined VoidCallback implementation that will be invoked when async create return.
+   *           It will contain the stats of the entry if succeeded.
+   */
   void asyncExist(final String key, AsyncCallback.StatCallback cb);
 
-  void asyncDelete(final String keys, AsyncCallback.VoidCallback cb);
+  /**
+   * The asynchronous version of delete.
+   * @param key key to identify the entry
+   * @param cb An user defined VoidCallback implementation that will be invoked when async delete return.
+   */
+  void asyncDelete(final String key, AsyncCallback.VoidCallback cb);
 
-  void asyncTransaction(final String keys, AsyncCallback.TransactionCallback cb);
+  /**
+   * The asynchronous version of transaction operations.
+   * @param ops A list of operations
+   * @param cb An user defined TransactionCallback implementation that will be invoked when transaction operations return.
+   */
+  void asyncTransaction(final Iterable<Op> ops, AsyncCallback.TransactionCallback cb);
```

Comment:
Especially for transaction, if any rollback operations required, do you expect the callback is the one user providing rollback operations? Or this callback is just designed for user return.

Reply:
Had an offline discussion with Junkai. TransactionCallback will be executed when the transactional operation finished, and the callback implementation should read the execution result return code and perform corresponding action. 
Will add more detail in comment section.